### PR TITLE
ARCHBOM-1584: refactor monitoring app

### PR DESCRIPTION
**Pull request by robrap**
**Description:**

* Add ADR to explain new app organization.
* Apply new app organization to the monitoring Django app.
* Remove middleware ordering checks. This is not a typical Django feature and it is painful when refactoring.
* Add CachedCustomMonitoringMiddleware, CodeOwnerMonitoringMiddleware, and MonitoringMemoryMiddleware to the public API.
* Deprecate above middleware from old location.
* Deprecate methods from modules that were always meant to be used from the public API:
  * accumulate
  * increment
  * set_custom_attribute
  * set_custom_attributes_for_course_key
* Add additional custom attributes for deprecated classes and methods to make them safer to retire.

Note: Some additional details in the CHANGELOG.rst change in the PR.

**JIRA:**

[ARCHBOM-1584](https://openedx.atlassian.net/browse/ARCHBOM-1584)

**Testing instructions:**

- [X] I will test with an edx-platform sandbox once edx-platform PR is ready.

**Merge checklist:**
- [x] Merge https://github.com/edx/edx-platform/pull/25504
- [X] All reviewers approved
- [x] CI build is green
- [X] Version bumped
- [X] Changelog record added
- [X] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

File: edx_django_utils/cache/middleware.py
Comment (robrap) [inline edx_django_utils/cache/middleware.py:6]: **Reviewer Note:** I apologize that I didn't remove `_check_middleware_dependencies` first in a separate PR, but I did it after refactor, and merge conflicts would have been painful.  It should be pretty obvious and easy to review.
Please mark as resolved once you've read this.

File: edx_django_utils/cache/tests/test_middleware.py
Comment (robrap) [inline edx_django_utils/cache/tests/test_middleware.py:18]: **Reviewer Note:** FYI: All of these `override_settings of `MIDDLEWARE` deletions are related to removing `_check_middleware_dependencies`.
Please mark as resolved once you've read this.

File: edx_django_utils/monitoring/internal/code_owner/middleware.py
Comment (robrap) [inline edx_django_utils/monitoring/internal/code_owner/middleware.py:2]: **Reviewer Note:** Outside of minor updates (e.g. import updates, minor doc changes involving imports, etc.), the logic of moved files should be mostly unchanged.
Please mark as resolved once you've read this.

File: edx_django_utils/monitoring/tests/code_owner/test_utils.py
Comment (robrap) [inline edx_django_utils/monitoring/tests/code_owner/test_utils.py:29]: **Reviewer Note:** FYI: I removed these tests of an internal method. Equivalent cases are already tested on other public calls that us this method.
Please mark as resolved once you've read this.

**Review by robrap — COMMENTED** [review]
Reviewer notes to help with the review.  Please resolve comments that you are done with.  Thank you.

**Review by feanil — APPROVED** [review]
Generally looks good, I just had a question on the removal of the

**Comment by robrap** [comment]
Thanks for your thoughts, @idegtiarov.  I'll look into whatever is reasonably quick. 
1. Black is hopefully coming in the future to help with code consistency.
2. I was trying to squeeze in a step in the right direction, but there will always be more that can be done, especially for existing code that is being moved around.

File: edx_django_utils/monitoring/internal/code_owner/middleware.py
Comment (robrap) [reply edx_django_utils/monitoring/internal/code_owner/middleware.py:58]: This code could definitely be refactored to be more clear, but I don't think it is invalid.  (In case it wasn't clear, this is not new code.  It is just being moved.)

As noted in the comment, the case is when the module is found in the request, but there wasn't a code owner configured for that path.  At that point, the only thing left to check is the catch all.  The transaction check lower down is for when the module wasn't found using the request.

I am going to be working in the code_owner code later (which is why I am doing this refactor), and I will consider looking into cleaning this up then.

File: edx_django_utils/monitoring/internal/code_owner/middleware.py
Comment (robrap) [reply edx_django_utils/monitoring/internal/code_owner/middleware.py:69]: Same answer.

File: edx_django_utils/monitoring/internal/code_owner/middleware.py
Comment (robrap) [reply edx_django_utils/monitoring/internal/code_owner/middleware.py:104]: Maybe this is a bad pattern, but it was put in place because I want to ensure that nothing can get through, even unplanned for errors, because I don't want to chance breaking a transaction.  I want to ensure this code is safe.

File: edx_django_utils/monitoring/internal/middleware.py
Comment (robrap) [reply edx_django_utils/monitoring/internal/middleware.py]: Note: there are future projects that will do this kind of clean-up.

**Review by robrap — COMMENTED** [review]
@idegtiarov: I did what made sense on this PR, and tried to add explanations.

I still have code coverage I need to add to land this PR.

**Comment by idegtiarov** [comment]
@robrap Thanks for the answers and clarifications.

File: edx_django_utils/monitoring/internal/code_owner/middleware.py
Comment (regisb) [reply edx_django_utils/monitoring/internal/code_owner/middleware.py:104]: Hey! :wave: Sorry to pitch in, but somehow I noticed this conversation in my inbox.

I think we must really, really avoid this pattern. I see a few major issues here:

1. The exception is cast as a string, which will prevent later raising an exception with the right type and the right error message. For instance, we won't be able to do `logging.exception(...)` anymore (which is very useful, e.g with sentry).
2. The Exception type is overly broad, and this will silence things such as SyntaxError, AttributeError, ImportError -- errors that we do want to learn about, whether it is in development or production. This [happens](https://github.com/edx/course-discovery/pull/2730).

I might not understand the full context here: I thought that the whole point of transactions was that we could interrupt them and be ok, because the entire transaction is rolled back on error?

Comment (robrap) [reply edx_django_utils/monitoring/internal/code_owner/middleware.py:104]: Thanks @regisb.  When this code was written, I very intentionally did **not** want this to interrupt or break any transactions.  At this time, we will be relying on this data more and more, so I should be able to fix this in a future PR when I updated some of the code_owner code.

**Comment by robrap** [comment]
@feanil: FYI: I plan on merging, but I wanted to let you know I added these tests for an existing middleware to increase coverage: https://github.com/edx/edx-django-utils/blob/241bb42ae5f511c25ab095018d05966cd509c63d/edx_django_utils/monitoring/tests/test_middleware.py

**Pull request merged**
